contest/scoreboard.py

```diff
@@ -0,0 +1,139 @@
+    copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+    The above copyright notice and this permission notice shall be included in all
+    copies or substantial portions of the Software.
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+    '''
+import sys
+from operator import methodcaller
+class Scoreboard:
+    def __init__(self,start_time):
+        self.users = []
+        self.problems = []
+        self.start_time = start_time
+
+    def add_user(self,user):
+        self.users.append(user)
+
+    def add_problem(self,scoreboard_problem):
+        self.problems.append(scoreboard_problem)
+
+    def get_problem(self,id):
+        for scoreboard_problem in self.problems:
+            if (scoreboard_problem.id == id):
+                return scoreboard_problem
+
+    #sort by solved descending. if same sort by penalty
+    def sort_users_by_penalty(self):
+        self.users = sorted(self.users, key=methodcaller('penalty',self.start_time))
+        self.users = sorted(self.users, key=methodcaller('solved'), reverse=True)
+
+    def sort_users_by_solved_testcases(self):
+        self.users = sorted(self.users, key=methodcaller('testcases_solved'), reverse=True)
+
+#for scoreboard
+class Scoreboard_Problem:
+    def __init__(self,id,pname,total_testcase):
+        self.id = id
+        self.pname = pname
+        self.total_testcase = total_testcase
+        self.pass_user = 0
+
+    def add_pass_user(self):
+        self.pass_user += 1
+
+class User:
+    def __init__(self,username):
+        self.username = username
+        self.problems = []
+
+    def add_problem(self,problem):
+        self.problems.append(problem)
+
+    def get_problem(self,id):
+        for problem in self.problems:
+            if(problem.id == id):
+                return problem
+
+    def solved(self):
+        count = 0
+        for problem in self.problems:
+            if(problem.is_solved() == True):
+                count += 1
+        return count
+
+    def testcases_solved(self):
+        count = 0
+        for problem in self.problems:
+            count += problem.testcases_solved()
+        return count
+
+    def penalty(self,start_time):
+        penalty = 0
+        for problem in self.problems:
+            penalty += problem.penalty(start_time);
+        return penalty
+
+#for each user
+class Problem:
+    def __init__(self,id):
+        self.submissions = []
+        self.id = id
+
+    def is_solved(self):
```

Comment:
but this will lead to redundant store of total testcases.
That is, every user has his own total testcases
I think of this while designing, I wonder which one is better